File: cases/taint/scala/closure-capture-positive/ClosureCapturePositive.scala
package dataflowbench

object ClosureCapturePositive {
  def dfb_source(): String = { // DFB-SOURCE: closure-capture-input
    "tainted"
  }

  def dfb_sink(value: String): Unit = {} // DFB-SINK: closure-capture-sink

  def makeReporter(): () => Unit = {
    val captured = dfb_source() // DFB-WITNESS: closure-capture-bind
    () => dfb_sink(captured)
  }

  def run(): Unit = {
    val reporter = makeReporter()
    reporter()
  }
}

File: cases/taint/kotlin/closure-capture-positive/ClosureCapturePositive.kt
package dataflowbench

object ClosureCapturePositive {
    fun dfb_source(): String { // DFB-SOURCE: closure-capture-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: closure-capture-sink

    fun makeReporter(): () -> Unit {
        val captured = dfb_source() // DFB-WITNESS: closure-capture-bind
        return { dfb_sink(captured) }
    }

    fun run() {
        val reporter = makeReporter()
        reporter()
    }
}
